beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -0,0 +1,226 @@
+    // while maintaining the invariant that block and data columns are persisted atomically.
+    let (data_columns_sender, data_columns_receiver) = tokio::sync::mpsc::channel(1);
+    let peer_das_enabled = chain.spec.is_peer_das_enabled_for_epoch(block.epoch());
+
+    // Partial blobs response isn't useful for PeerDAS, so we don't bother building and publishing data columns.
+    if peer_das_enabled && all_blobs_fetched {
+        let logger = chain.log.clone();
+        let block_cloned = block.clone();
+        let kzg = chain.kzg.clone().expect("KZG not initialized");
+        let spec = chain.spec.clone();
+        let blobs_cloned = fixed_blob_sidecar_list.clone();
+        let chain_cloned = chain.clone();
+        chain
+            .task_executor
+            .spawn_handle(
+                async move {
+                    let blob_refs = blobs_cloned.iter()
+                        .filter_map(|b| b.as_ref().map(|b| &b.blob))
+                        .collect::<Vec<_>>();
+                    let data_columns_result = DataColumnSidecar::build_sidecars(
+                        &blob_refs,
+                        &block_cloned,
+                        &kzg,
+                        &spec,
+                    );
+
+                    let all_data_columns = match data_columns_result {
+                        Ok(d) => d,
+                        Err(e) => {
+                            error!(logger, "Failed to build data column sidecars from blobs"; "error" => ?e);
+                            return;
+                        }
+                    };
+
+                    // Check indices from cache before sending the columns, to make sure we don't
+                    // publish components already seen on gossip.
+                    let all_data_columns_iter = all_data_columns.clone().into_iter();
+                    let data_columns_to_publish = match chain_cloned.data_availability_checker.imported_custody_column_indexes(&block_root) {
+                        None => Either::Left(all_data_columns_iter),
+                        Some(imported_columns_indices) => Either::Right(all_data_columns_iter.filter(move |d| !imported_columns_indices.contains(&d.index()))),
+                    }.collect::<Vec<_>>();
+
+                    if let Err(e) = data_columns_sender.try_send(all_data_columns) {
+                        error!(logger, "Failed to send computed data columns"; "error" => ?e);
+                    };
+
+                    let is_supernode =
+                        chain_cloned.data_availability_checker.get_custody_columns_count() == spec.number_of_columns;
+                    if is_supernode && !data_columns_to_publish.is_empty() {
+                        publish_fn(BlobsOrDataColumns::DataColumns(data_columns_to_publish));
+                    }
+                },
+                "compute_data_columns",
+            )
+            .ok_or(BeaconChainError::RuntimeShutdown)?;
+    } else {
```

Comment:
This else will attempt to publish blobs on peerdas

Reply:
Fixed